server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearchTest.java

```diff
@@ -173,16 +186,209 @@ public void testGetSpanFromEmptyDB() throws StoreException, InterruptedException
+        Assert.assertThat(descendantProducerNode, instanceOf(Producer.class));
+        Assert.assertEquals(1, descendantProducerNode.getCorrelationIds().size());
+        Assert.assertEquals("descendant", descendantProducerNode.getCorrelationIds().get(0).getValue());
+        Assert.assertTrue(descendantProducerNode.getNodes().isEmpty());
+
+        InteractionNode descendantProducerNode2 = ((InteractionNode) rootConsumerNode.getNodes().get(1));
+        Assert.assertThat(descendantProducerNode2, instanceOf(Producer.class));
+        Assert.assertEquals(1, descendantProducerNode2.getCorrelationIds().size());
+        Assert.assertEquals("descendant2", descendantProducerNode2.getCorrelationIds().get(0).getValue());
+        Assert.assertTrue(descendantProducerNode2.getNodes().isEmpty());
+    }
+
+    @Test
+    public void testGetTraceFragmentClientRoot() throws StoreException, InterruptedException {
+        Span rootClientSpan = new Span();
+        rootClientSpan.setId("root");
+        rootClientSpan.setTraceId("root");
+        rootClientSpan.setName("rootClientSpan");
+        rootClientSpan.setAnnotations(clientAnnotations());
+
+        Span serverSpan = new Span();
+        serverSpan.setId("root");
+        serverSpan.setTraceId("root");
+        serverSpan.setAnnotations(serverAnnotations());
+
+        Span clientSpanDescendantRoot = new Span();
+        serverSpan.setId("root2");
+        serverSpan.setParentId("root");
+        serverSpan.setTraceId("root");
+        serverSpan.setAnnotations(clientAnnotations());
+
+        Span serverSpanOfClientDescendantRoot = new Span();
+        serverSpanOfClientDescendantRoot.setId("root2");
+        serverSpanOfClientDescendantRoot.setParentId("root");
+        serverSpanOfClientDescendantRoot.setTraceId("root");
+        serverSpanOfClientDescendantRoot.setAnnotations(serverAnnotations());
+
+        storeAndWait(null, Arrays.asList(
+                rootClientSpan,
+                serverSpan,
+                clientSpanDescendantRoot,
+                serverSpanOfClientDescendantRoot),
+                SpanUniqueIdGenerator::toUnique);
+
+        Trace trace = spanService.getTraceFragment(null, SpanUniqueIdGenerator.getClientId("root"));
+        Assert.assertEquals("root", trace.getId());
+        Assert.assertEquals(1, trace.getNodes().size());
+
+        InteractionNode rootProducerNode = (InteractionNode) trace.getNodes().get(0);
+        Assert.assertThat(rootProducerNode, instanceOf(Producer.class));
+        Assert.assertEquals(1, rootProducerNode.getCorrelationIds().size());
+        Assert.assertEquals("root", rootProducerNode.getCorrelationIds().get(0).getValue());
+        Assert.assertEquals(1, rootProducerNode.getNodes().size());
+        Assert.assertEquals(new HashSet<>(Arrays.asList("root2")),
+                extractCorrelationIds(rootProducerNode.getNodes()));
+
+        InteractionNode descendantConsumerNode = (InteractionNode) rootProducerNode.getNodes().get(0);
```

Comment:
It shouldn't be possible for a node to be contained in a Producer (not in a trace fragment anyway), so something must have gone wrong here.